Allow fetching or deleting data from custom persistent credential storage

szewai opened the pull request:
#### 1eb5f68056d281580af0392b1d3cc1ee7c79b20e
<pre>
Allow fetching or deleting data from custom persistent credential storage
<a href="https://bugs.webkit.org/show_bug.cgi?id=252126">https://bugs.webkit.org/show_bug.cgi?id=252126</a>
rdar://105347284

Reviewed by Alex Christensen.

Since 260094@main, WebsiteDataStore created with identifier uses separate credential storage from the shared (default)
one. Therefore, network process should look at correct credential storage when doing data fetch/deletion. For
WebsiteDataStore using shared credential storage, the current behavior is to only fetch/delete session-only credentials.
For custom (separate) credential storage, which should only be used by WebKit clients that has access to the
WebsiteDataStore, we could handle all credentails.

* Source/WebCore/platform/network/CredentialStorage.cpp:
(WebCore::CredentialStorage::originsWithSessionCredentials): Deleted.
(WebCore::CredentialStorage::removeSessionCredentialsWithOrigins): Deleted.
(WebCore::CredentialStorage::clearSessionCredentials): Deleted.
* Source/WebCore/platform/network/CredentialStorage.h:
* Source/WebCore/platform/network/mac/CredentialStorageMac.mm:
(WebCore::CredentialStorage::originsWithSessionCredentials): Deleted.
(WebCore::CredentialStorage::removeSessionCredentialsWithOrigins): Deleted.
(WebCore::CredentialStorage::clearSessionCredentials): Deleted.
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::fetchWebsiteData):
(WebKit::NetworkProcess::deleteWebsiteData):
(WebKit::NetworkProcess::deleteWebsiteDataForOrigins):
(WebKit::NetworkProcess::deleteAndRestrictWebsiteDataForRegistrableDomains):
(WebKit::NetworkProcess::registrableDomainsWithWebsiteData):
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::originsWithCredentials):
(WebKit::NetworkSessionCocoa::removeCredentialsForOrigins):
(WebKit::NetworkSessionCocoa::clearCredentials):

Canonical link: <a href="https://commits.webkit.org/260276@main">https://commits.webkit.org/260276@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4ace5727254bcad060aa0981a1964e63122c91e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107729 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16791 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40633 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116882 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116265 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111619 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18206 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8107 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99911 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113483 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13747 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96946 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41434 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95655 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28586 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9758 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29934 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10439 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6823 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15912 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49518 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7109 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12001 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
```diff
@@ -1690,6 +1690,72 @@ static void activateSessionCleanup(NetworkSessionCocoa& session, const NetworkSe
+    bool shouldHandleSessionCredentialsOnly = credentialStorage == [NSURLCredentialStorage sharedCredentialStorage];
+    auto allCredentials = [credentialStorage allCredentials];
+    HashSet<WebCore::SecurityOriginData> origins;
+    for (NSURLProtectionSpace *space in allCredentials) {
```
It may be slightly more efficient to iterate allCredentials.allValues instead of doing hash lookups on each one.

achristensen07 commented on Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
```diff
@@ -1690,6 +1690,72 @@ static void activateSessionCleanup(NetworkSessionCocoa& session, const NetworkSe
+
+    bool shouldHandleSessionCredentialsOnly = credentialStorage == [NSURLCredentialStorage sharedCredentialStorage];
+    auto allCredentials = [credentialStorage allCredentials];
+    for (NSURLProtectionSpace *space in allCredentials) {
```
ditto?  Also, allCredentials doesn't need to be stored separately.

szewai committed:
Allow fetching or deleting data from custom persistent credential storage
https://bugs.webkit.org/show_bug.cgi?id=252126
rdar://105347284

Reviewed by Alex Christensen.

Since 260094@main, WebsiteDataStore created with identifier uses separate credential storage from the shared (default)
one. Therefore, network process should look at correct credential storage when doing data fetch/deletion. For
WebsiteDataStore using shared credential storage, the current behavior is to only fetch/delete session-only credentials.
For custom (separate) credential storage, which should only be used by WebKit clients that has access to the
WebsiteDataStore, we could handle all credentails.

* Source/WebCore/platform/network/CredentialStorage.cpp:
(WebCore::CredentialStorage::originsWithSessionCredentials): Deleted.
(WebCore::CredentialStorage::removeSessionCredentialsWithOrigins): Deleted.
(WebCore::CredentialStorage::clearSessionCredentials): Deleted.
* Source/WebCore/platform/network/CredentialStorage.h:
* Source/WebCore/platform/network/mac/CredentialStorageMac.mm:
(WebCore::CredentialStorage::originsWithSessionCredentials): Deleted.
(WebCore::CredentialStorage::removeSessionCredentialsWithOrigins): Deleted.
(WebCore::CredentialStorage::clearSessionCredentials): Deleted.
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::fetchWebsiteData):
(WebKit::NetworkProcess::deleteWebsiteData):
(WebKit::NetworkProcess::deleteWebsiteDataForOrigins):
(WebKit::NetworkProcess::deleteAndRestrictWebsiteDataForRegistrableDomains):
(WebKit::NetworkProcess::registrableDomainsWithWebsiteData):
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::originsWithCredentials):
(WebKit::NetworkSessionCocoa::removeCredentialsForOrigins):
(WebKit::NetworkSessionCocoa::clearCredentials):

Canonical link: https://commits.webkit.org/260276@main
